filebeat/input/file/identifier.go

```diff
@@ -0,0 +1,219 @@
+	return &inodeDeviceIdentifier{
+		name: inodeDeviceIDName,
+	}, nil
+}
+
+func (i *inodeDeviceIdentifier) GenerateID(s *State) {
+	stateID := i.name + identitySep + s.FileStateOS.String()
+	if len(s.Meta) == 0 {
+		s.Id = stateID
+	} else {
+		s.Id = genIDWithHash(s.Meta, stateID)
+	}
+	s.IdentifierName = i.name
+}
+
+func (i *inodeDeviceIdentifier) Name() string {
+	return i.name
+}
+
+type pathIdentifier struct {
+	name string
+}
+
+func newPathIdentifier(_ *common.Config) (StateIdentifier, error) {
+	return &pathIdentifier{
+		name: pathName,
+	}, nil
+}
+
+func (p *pathIdentifier) GenerateID(s *State) {
+	stateID := p.name + identitySep + s.Source
+	if len(s.Meta) == 0 {
+		s.Id = stateID
+	} else {
+		s.Id = genIDWithHash(s.Meta, stateID)
+	}
+	s.IdentifierName = p.name
+}
+
+func (p *pathIdentifier) Name() string {
+	return p.name
+}
+
+type inodeMarkerIdentifier struct {
+	log        *logp.Logger
+	name       string
+	markerPath string
+
+	markerFileLastModifitaion time.Time
+	markerTxt                 string
+}
+
+func newINodeMarkerIdentifier(cfg *common.Config) (StateIdentifier, error) {
+	var config struct {
+		MarkerPath string `config:"path" validate:"required"`
+	}
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("error while reading configuration of INode + marker file configuration: %v", err)
+	}
+
+	fi, err := os.Stat(config.MarkerPath)
+	if err != nil {
+		return nil, fmt.Errorf("error while opening marker file at %s: %v", config.MarkerPath, err)
+	}
+	markerContent, err := ioutil.ReadFile(config.MarkerPath)
+	if err != nil {
+		return nil, fmt.Errorf("error while reading marker file at %s: %v", config.MarkerPath, err)
+	}
+	return &inodeMarkerIdentifier{
+		log:                       logp.NewLogger("inode_marker_identifier_" + filepath.Base(config.MarkerPath)),
+		name:                      inodeMarkerName,
+		markerPath:                config.MarkerPath,
+		markerFileLastModifitaion: fi.ModTime(),
+		markerTxt:                 string(markerContent),
+	}, nil
+}
+
+func (i *inodeMarkerIdentifier) markerContents() string {
+	f, err := os.Open(i.markerPath)
+	if err != nil {
+		i.log.Errorf("Failed to open marker file %s: %v", i.markerPath, err)
+		return ""
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		i.log.Errorf("Failed to fetch file information for %s: %v", i.markerPath, err)
+		return ""
+	}
+	if i.markerFileLastModifitaion.Before(fi.ModTime()) {
+		contents, err := ioutil.ReadFile(i.markerPath)
+		if err != nil {
+			i.log.Errorf("Error while reading contents of marker file: %v", err)
+			return ""
+		}
+		i.markerTxt = string(contents)
+	}
+
+	return i.markerTxt
+}
+
+func (i *inodeMarkerIdentifier) GenerateID(s *State) {
+	m := i.markerContents()
+
+	fileID := fmt.Sprintf("%s%s%d-%s", i.name, identitySep, s.FileStateOS.Inode, m)
+	if len(s.Meta) == 0 {
+		s.Id = fileID
+	} else {
+		s.Id = genIDWithHash(s.Meta, fileID)
+	}
```

Comment:
this `if len(s.Meta) == 0 { ... } else { ... } ` repeats itself all the time. Let's put it into a function (`genIDWithHash` could do the check).